sys/kern/signal.c

```diff
@@ -201,6 +203,81 @@ int do_sigsuspend(proc_t *p, const sigset_t *mask) {
   return ERESTARTNOHAND;
 }
 
+int do_sigtimedwait(proc_t *p, sigset_t waitset, ksiginfo_t *kinfo,
+                    timespec_t *tsp) {
+  int timevalid = 0, error = 0, timeout = 0;
+  sigset_t saved_mask;
+  signo_t sig;
+  thread_t *td = p->p_thread;
+
+  if (tsp != NULL) {
+    if (tsp->tv_nsec >= 0 && tsp->tv_nsec < 1000000000) {
+      timevalid = 1;
+      timeout = ts2hz(tsp);
+      /* Timeout set to 0 means to not block, but for sleepq it means to
+       * wait indefinitely */
+      if (timeout == 0) {
+        if (tsp->tv_nsec == 0 && tsp->tv_sec == 0) {
+          timeout = -1;
+        } else {
+          /* Shortest possible timeout */
+          timeout = 1;
+        }
+      }
+    }
+  }
+
+  bzero(kinfo, sizeof(*kinfo));
+
+  /* Silently ignore SIGKILL and SIGSTOP. */
+  __sigminusset(&cantmask, &waitset);
+
+  WITH_PROC_LOCK(p) {
+    /* Unblocking temporarly waited signals so we're woken up upon receiving
+     * such signal. */
+    saved_mask = td->td_sigmask;
+    __sigminusset(&waitset, &td->td_sigmask);
+
+    if ((sig = sig_pending(td))) {
+      sigpend_get(&td->td_sigpend, sig, kinfo);
+      error = 0;
+      goto out;
+    }
+    if (timeout == -1) {
+      error = EAGAIN;
+      goto out;
+    }
+    if (tsp != NULL && !timevalid) {
+      error = EINVAL;
+      goto out;
+    }
+  }
+
+  error = sleepq_wait_timed(&td->td_sigmask, "sigtimedwait()", timeout);
```

Comment:
Please use here new interface for `sleepq_wait_timed` and pass `&p->p_lock` to it.